### PR TITLE
Jvenstad/pin versions

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden please review.

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/Upgrader.java [hunk 1]
```diff
@@ -92,8 +92,10 @@ public void maintain() {
         return reversed;
     }
 
-    /** Returns a list of all applications */
-    private ApplicationList applications() { return ApplicationList.from(controller().applications().asList()); }
+    /** Returns a list of all applications, except those which are pinned — these should not be manipulated by the Upgrader */
+    private ApplicationList applications() {
+        return ApplicationList.from(controller().applications().asList()).notPinned();
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/Upgrader.java]: Consider moving all filtering to the `upgrade` method, together with the others. Easier to read when all filtering happens in one place.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/Change.java [hunk 2]
```diff
@@ -57,17 +60,32 @@ public boolean isPresent() {
     /** Returns the application version carried by this. */
     public Optional<ApplicationVersion> application() { return application; }
 
+    public boolean isPinning() { return pinning; }
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/Change.java]: Nit-pick: Shouldn't this be called `isPinned`?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializer.java [hunk 3]
```diff
@@ -275,13 +276,15 @@ private void jobRunToSlime(JobStatus.JobRun jobRun, Cursor parent, String jobRun
     }
 
     private void toSlime(Change deploying, Cursor parentObject, String fieldName) {
-        if ( ! deploying.isPresent()) return;
+        if ( ! deploying.isPresent() && ! deploying.isPinning()) return;
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializer.java]: Can `Change#isPresent` be updated to return `true` if pin is set? Or in other words, can a change only be considered empty if it's unpinned and platform and app version are unset?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/Upgrader.java (hunk 1 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/Upgrader.java]: It's not relevant just for upgrades, but also cancellation. That's why I put it here.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/Change.java (hunk 2 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/Change.java]: Not sure. I considered both. If you prefer isPinned, I'll let that sway my opinion.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializer.java (hunk 3 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializer.java]: That doesn't fit well with the usage (and doc) for `Change#isPresent`.

**Review by mpolden — APPROVED** [review]
Great!

**Pull request merged**